### PR TITLE
Implement the convenience methods count and collect in DataSet

### PR DESCRIPTION
**Pull request by mxm**
These methods provide convenience for the API user to get intermediate results into the program.

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/api/common/accumulators/ListAccumulator.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+	@Override
+	public void merge(Accumulator<T, List<T>> other) {
+		localValue.addAll(((ListAccumulator<T>) other).getLocalValue());
+	}
+
+	@Override
+	public void write(DataOutputView out) throws IOException {
+		ObjectOutputStream outStream = new ObjectOutputStream(new DataOutputViewStream(out));
+		outStream.writeObject(typeSerializer);
+		outStream.flush();
```
Comment (StephanEwen) [inline flink-core/src/main/java/org/apache/flink/api/common/accumulators/ListAccumulator.java]: Let us change this to `close()`

File: flink-java/src/main/java/org/apache/flink/api/java/DataSet.java [hunk 2]
```diff
@@ -302,7 +308,48 @@ public ExecutionEnvironment getExecutionEnvironment() {
...
-	
+
+	/**
+	 * Convenience method to get the count (number of elements) of a DataSet
+	 *
+	 * @return A long integer that represents the number of elements in the set
+	 *
+	 * @see org.apache.flink.api.java.Utils.CountHelper
+	 */
+	public Long count() throws Exception {
```
Comment (StephanEwen) [inline flink-java/src/main/java/org/apache/flink/api/java/DataSet.java]: I would make this a primitive `long`

File: flink-java/src/main/java/org/apache/flink/api/java/Utils.java [hunk 3]
```diff
@@ -32,4 +38,44 @@ public static String getCallLocationName(int depth) {
...
+
+		private final String id;
+
+		public CountHelper(String id) {
+			this.id = id;
+		}
+
+		@Override
+		public void flatMap(T value, Collector<Long> out) throws Exception {
+			getRuntimeContext().getLongCounter(id).add(1L);
```
Comment (StephanEwen) [inline flink-java/src/main/java/org/apache/flink/api/java/Utils.java]: I think it is more efficient to have a `long`inside this function that is incremented for each record, and then aggregate this `long` to the accumulator in the `close()` method.

File: flink-runtime/src/test/java/org/apache/flink/runtime/AbstractIDTest.java [hunk 4]
```diff
@@ -23,8 +23,8 @@
 import static org.junit.Assert.fail;
 
 import org.junit.Test;
-
 import org.apache.flink.runtime.testutils.CommonTestUtils;
+import org.apache.flink.util.AbstractID;
```
Comment (StephanEwen) [inline flink-runtime/src/test/java/org/apache/flink/runtime/AbstractIDTest.java]: Let us move this test to `flink-core`, with `AbstractID` being now part of the core.

**Comment by StephanEwen** [comment]
With the scheduler and intermediate data set enhancements coming up for 0.9 soon, this is now quite feasible to use. I suggest to merge it once the inline comments are addressed.

**Comment by mxm** [comment]
I fixed the proposed changes and rebased to the current master. The changes are only reflected in the Java API and need to be added to the Scala API as well.

File: flink-runtime/src/test/java/org/apache/flink/runtime/AbstractIDTest.java (hunk 4 above)
Comment (mxm) [inline flink-runtime/src/test/java/org/apache/flink/runtime/AbstractIDTest.java]: As far as I see, `AbstractIDTest` is already part of `flink-core`.

Comment (mxm) [inline flink-runtime/src/test/java/org/apache/flink/runtime/AbstractIDTest.java]: Nevermind, I forgot to push the last changes I made..

**Comment by mxm** [comment]
I've implemented count and collect in the Scala API. There is still a problem with the `ListAccumulator` for non-primitive Objects (e.g. not Integer or Long) probably due to Object reuse.

**Comment by mxm** [comment]
Looks like this is now ready to merge. 

@zentol I understand your concern. However, I think that it is much easier to execute in this way. Most of the times, the user probably wants just one accumulator result and not multiple. This is supposed to be a convenience function.

**Comment by StephanEwen** [comment]
@zentol  You are right, for the time being, that this results in parts in repeated execution. While not totally unavoidable in all cases, the code going in soon about caching intermediate results will help there big time.

**Comment by mxm** [comment]
I squashed the commits and rebased to the current master. Any objections against merging this?

File: flink-core/pom.xml [hunk 5]
```diff
@@ -58,6 +58,12 @@ under the License.
 			<version>0.5.1</version>
 		</dependency>
 
+		<dependency>
```
Comment (rmetzger) [inline flink-core/pom.xml]: Why are you moving netty to `flink-core` ?

Comment (mxm) [inline flink-core/pom.xml]: Simply because AbstractID depends on `io.netty.buffer.ByteBuf`

Comment (rmetzger) [inline flink-core/pom.xml]: Allright, that makes sense

Comment (uce) [inline flink-core/pom.xml]: Actually, only a single convenience method in `AbstractID` depends on this. I would propose to refactor that method into one of the existing utility classes in `flink-runtime` and keep the Netty dependency in `flink-runtime`.

**Comment by StephanEwen** [comment]
Let us not move netty to flink-core. Let us rather pull that method out of the AbstractID, or, not use AbstractID at in your code, but the Java UUID.

**Comment by mxm** [comment]
I moved the `writeInto(ByteBuf buf)` method to the subclasses of `AbstractID` used by `NettyMessage`. This is the only context where the method is being used and the subclasses already have another `fromByteBuf(ByteBuf buf)` method.

**Comment by mxm** [comment]
If there are no objections, I will merge this once Travis passes.

**Pull request closed without merging**